win32ss/user/ntuser/window.c

```diff
@@ -3996,19 +3997,33 @@ NtUserQueryWindow(HWND hWnd, DWORD Index)
          Result = (pWnd->head.pti->MessageQueue == gpqForeground);
          break;
 
-      case QUERY_WINDOW_DEFAULT_IME:
-         ERR("QUERY_WINDOW_DEFAULT_IME: FIXME\n");
-         Result = 0;
+      case QUERY_WINDOW_DEFAULT_IME: /* default IME window */
+         if (pWnd->head.pti->spwndDefaultIme)
+            Result = (DWORD_PTR)UserHMGetHandle(pWnd->head.pti->spwndDefaultIme);
+         else
+            Result = 0;
          break;
 
-      case QUERY_WINDOW_DEFAULT_ICONTEXT:
-         ERR("QUERY_WINDOW_DEFAULT_ICONTEXT: FIXME\n");
-         Result = 0;
+      case QUERY_WINDOW_DEFAULT_ICONTEXT: /* default input context handle */
+         if (pWnd->head.pti->spDefaultImc)
+            Result = (DWORD_PTR)UserHMGetHandle(pWnd->head.pti->spDefaultImc);
+         else
+            Result = 0;
          break;
 
       case QUERY_WINDOW_ACTIVE_IME:
-         ERR("QUERY_WINDOW_ACTIVE_IME: FIXME\n");
          Result = 0;
+         if (gpqForeground && gpqForeground->spwndActive)
+         {
+             pwndActive = gpqForeground->spwndActive;
+             pti = PsGetCurrentThreadWin32Thread();
+             if (pwndActive && pti->rpdesk == pwndActive->head.rpdesk)
```

Comment:
```suggestion
             if (pti && pti->rpdesk == pwndActive->head.rpdesk)
```

Reply:
And no check of whether `pwndActive` is null?

Reply:
Sir, what's your command?

I thought `pwndActive` null check was necessary due to multithread. Shall I add `pti` null check?

Reply:
@katahiromz Do as it is done in the other places of the win32k code where `spwndActive` is being used.

Reply:
> Shall I add `pti` null check?

I assumed there was a typo, but, if `pti` null check is useless here, then do not add it either.

Reply:
> > no check of whether `pwndActive` is null?
> 
> Previous `if` already checks that, doesn't it?
> 

Hehe thanks, indeed it does it (the check that is done together with `gpqForeground`)

> > multithread
> 
> If value(s) can change, then I guess that is a bigger problem...

win32k is single-threaded in the sense that, as we can see, all the exported functions are protected with the global lock (see e.g. `UserEnterShared()` etc.), so that if two(or more) separate user-mode threads call win32k functions at the "same time", all except one call gets to wait its own turn.
And for internal win32k functions that run by themselves (w/o being called from user-mode), if they are designed correctly, they should also lock whenever they are going to access a global, or some data that belongs to an exposed object (like windows, desktops, etc.)

Here, `NtUserQueryWindow()` satisfies this: it uses the global lock (see the `UserEnterShared()` call) and therefore this should protect e.g. `gpqForeground` and its data (e.g. `gpqForeground->spwndActive`), plus the other global/user-exposed stuff touched there, to be modified under the hood.
So for example, the capture of the `spwndActive` into `ptiActive` is not necessary (but it just makes the lines of code later being shorter, which is nice)...

Reply:
Regarding the `pti` NULL check, this is redundant. Indeed, **any** win32 thread that calls **any** win32k-exported function (the `NtUser...` etc. ones), are automatically transformed into a Win32 thread, from the point of view of win32k, before proceeding further. This involves the creation + initialization of what is going to be returned by `PsGetCurrentThreadWin32Thread`,  **if it doesn't exist already**, so that, in win32k, when being called within any `NtUser...` function, `PsGetCurrentThreadWin32Thread` **always** returns a valid (and not NULL) pointer to a valid block of data.

Reply:
@HBelusca, thanks for the detailed explanations.

@katahiromz, simply remove the extra `pwndActive` check.

Reply:
Ok.